Fix bigbird random attention

Bearnardd opened the pull request:
# What does this PR do?

Fixes the bug mentioned in the [issue](https://github.com/huggingface/transformers/issues/17355) by transiting from `np.random` to the `jax.random`. It also adds several minor changes to be able to run the new code and pass the all the tests 

<!-- Remove if not applicable -->

Fixes # (issue)
https://github.com/huggingface/transformers/issues/17355


## Before submitting

- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.

## Who can review?

@sanchit-gandhi @thevasudevgupta @patrickvonplaten

Bearnardd committed:
switch np.random.permutation to jax.random.permuation

Bearnardd committed:
Merge branch 'main' of https://github.com/huggingface/transformers into fix_bigbird_random_attention

Bearnardd committed:
remove comments

Bearnardd committed:
remove leftover comment

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21023). All of your documentation changes will be reflected on that endpoint.

Bearnardd committed:
skip similarity tests

Bearnardd committed:
Merge branch 'fix_bigbird_random_attention' of https://github.com/Bearnardd/transformers into fix_bigbird_random_attention

patrickvonplaten commented on tests/test_modeling_flax_common.py:
```diff
@@ -133,7 +133,7 @@ def _prepare_for_class(self, inputs_dict, model_class):
             inputs_dict = {
                 k: jnp.broadcast_to(v[:, None], (v.shape[0], self.model_tester.num_choices, v.shape[-1]))
-                if isinstance(v, (jnp.ndarray, np.ndarray))
+                if isinstance(v, (jnp.ndarray, np.ndarray)) and k != "indices_prng_key"
```
Ok for me!

patrickvonplaten commented on tests/models/big_bird/test_modeling_flax_big_bird.py:
```diff
@@ -221,3 +227,15 @@ def check_pt_flax_outputs(self, fx_outputs, pt_outputs, model_class, tol=1e-5, n
+    @is_pt_flax_cross_test
+    # due to jit compilation flax version has to use jax.random with
+    # provided PRGN key instead of np.random
+    def test_equivalence_flax_to_pt(self):
```
Is there any way we could make these tests pass instead of removing them? Or write a new test?

Bearnardd replied:
@patrickvonplaten I was not able to think about any thing in particular yet. But I will give it a try during a weekend.

sanchit-gandhi replied:
Good catch @amyeroberts! If we set `train=False` here for the Flax model and `model.eval()` for the PyTorch one then there should be no discrepancy (since we won't use any randomness), worth double checking the parent tester class:
https://github.com/huggingface/transformers/blob/d00997e66c0f4f6e55a1a422d6d3ca8fb1d37aab/tests/test_modeling_flax_common.py#L350

Bearnardd replied:
Hi @sanchit-gandhi @amyeroberts! I had some time to look at the equivalence tests and either I am missing something really simple or there is some missing part in the pytorch implementation of `BigBird` since setting it up to `eval` mode does not reflect in any change in its behavior corresponding to the random attention. What I mean by that is that I cannot find any part in the code that controls whether to use random attentions or not depending on whether we are in the `eval` or `train` mode. It seems like it is always using random attention no matter the mode. Moreover if I manually overwrite the rand_attn in pytorch implementation with `np.zeros` then the outputs from `flax` and `pytorch` match.

Bearnardd replied:
@sanchit-gandhi that is a great idea! I will skip the tests for now, open issue for the Pytorch version, fix it and "unskip" equivalence tests with that PR as well.

patrickvonplaten commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -240,7 +240,6 @@ def __call__(self, input_ids, token_type_ids, position_ids, attention_mask, dete
         return hidden_states
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertSelfAttention with Bert->BigBird
```
The changes in this file looks reasonable!

patrickvonplaten reviewed (commented):
Thanks a lot for your effort here @Bearnardd !

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1179,14 +1193,14 @@ def _get_single_block_row_attention(
             illegal_blocks.append(1)
 
-        selected_random_blokcs = []
+        selected_random_blocks = []
```
Thanks for the typo fix!

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1597,14 +1616,14 @@ def enable_gradient_checkpointing(self):
         position_ids = jnp.broadcast_to(jnp.arange(jnp.atleast_2d(input_ids).shape[-1]), input_shape)
         attention_mask = jnp.ones_like(input_ids)
         head_mask = jnp.ones((self.config.num_hidden_layers, self.config.num_attention_heads))
+        indices_prng_key = jax.random.PRNGKey(0)
 
         params_rng, dropout_rng = jax.random.split(rng)
         rngs = {"params": params_rng, "dropout": dropout_rng}
```
Note that here we set the PRNG key for the indices to 0 each time. This means that the PRNG key for the indices is fixed.

We should **never** re-use PRNG keys in JAX. Doing so gives identical outputs, thus destroying the stochastic nature of the random permutation function. See https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html#random-numbers-in-jax for more details.

Instead, we should generate new PRNG keys by _splitting_ the keys ourselves:

```suggestion
        rng, params_rng, dropout_rng, indices_rng = jax.random.split(rng, num=3)
        rngs = {"params": params_rng, "dropout": dropout_rng, "incides": indices_rng}
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1654,21 +1680,28 @@ def init_cache(self, batch_size, max_length):
         input_ids = jnp.ones((batch_size, max_length), dtype="i4")
         attention_mask = jnp.ones_like(input_ids, dtype="i4")
         position_ids = jnp.broadcast_to(jnp.arange(jnp.atleast_2d(input_ids).shape[-1]), input_ids.shape)
+        indices_prng_key = jax.random.PRNGKey(0)
```
Same again here: there should be no need to generate the PRNG key each time.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1698,6 +1731,9 @@ def __call__(
         if head_mask is None:
             head_mask = jnp.ones((self.config.num_hidden_layers, self.config.num_attention_heads))
 
+        if indices_prng_key is None:
+            indices_prng_key = jax.random.PRNGKey(0)
```
We shouldn't default the PRNG key to 0. Let's use the same logic that we employ for `dropout` and create an RNG that we assign to `indices`.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1698,6 +1731,9 @@ def __call__(
+            indices_prng_key = jax.random.PRNGKey(0)
+
         # Handle any PRNG if needed
         rngs = {}
```
This is how we should pass the indices PRNG key. Not as an argument, but rather in the dict of RNGs. This is the same logic that we use for dropout (see below):

```suggestion
        rngs = {}
        if indices_prng_key is not None:
            rngs["indices"] = indices_prng_key
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1722,6 +1758,7 @@ def __call__(
                 token_type_ids=jnp.array(token_type_ids, dtype="i4"),
                 position_ids=jnp.array(position_ids, dtype="i4"),
                 head_mask=jnp.array(head_mask, dtype="i4"),
+                indices_prng_key=jnp.array(indices_prng_key, dtype="u4"),
```
We shouldn't pass the indices PRNG key as an arg, but rather in the dict of RNGs (see above)

```suggestion
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1654,21 +1680,28 @@ def init_cache(self, batch_size, max_length):
         token_type_ids=None,
         position_ids=None,
         head_mask=None,
+        indices_prng_key=None,
```
(nit)
```suggestion
        indices_prng_key: jax.random.PRNGKey = None,
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1747,6 +1784,7 @@ def __call__(
                 jnp.array(input_ids, dtype="i4"),
                 jnp.array(attention_mask, dtype="i4"),
                 token_type_ids=jnp.array(token_type_ids, dtype="i4"),
+                indices_prng_key=jnp.array(indices_prng_key, dtype="u4"),
```
Again, we shouldn't pass this as an arg, but rather in the dict of RNGs.
```suggestion
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -447,6 +446,7 @@ def __call__(
         self,
         hidden_states,
         attention_mask,
+        indices_prng_key,
```
Let's not pass the PRNG key as an argument to the self attention layer. Instead, what we can do is copy the logic that we employ for dropout and 'make' an RNG variable:
https://github.com/huggingface/transformers/blob/b210c83a78022226ce48402cd67d8c8da7afbd8d/src/transformers/models/big_bird/modeling_flax_big_bird.py#L388-L390

For the indices PRNG key, this will look something as follows:
```python
 indices_prng_key = None 
 if not deterministic: 
     indices_prng_key = self.make_rng("indices")
``` 

We can then forward the variable `indices_prng_key` to the relevant sparse attention functions.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1233,6 +1247,7 @@ def __call__(
         self,
         hidden_states,
         attention_mask,
+        indices_prng_key,
```
We should remove the `indices_prng_key` argument from here onwards. It should not be passed as an arg to the self-attention layer. If you've got the rng dict approach correctly working, you should be able to remove all `indices_prng_key` args from here to the end of the modelling file!

```suggestion
```

sanchit-gandhi reviewed (commented):
Hey @Bearnardd! Thanks for the PR - we certainly do need to change all the `np.random` calls to `jnp.random` ones to correctly JIT the Flax model. The changes to using `jnp.random.permutation` in the self-attention module are on the right lines!

A few points from me:
1. Let's make sure that we get the correct ordering of the args in the `jax.random.permutation` function. According to the [docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.permutation.html), they should be (`key`, `array`) in that order.
2. We should not have to create the `indices_prng_key` each time. Doing so makes the JAX permutation function deterministic (undesirable behaviour). Instead, we should assign an RNG called `indices` in the attention function and pass the RNG key in our dict of RNGs (as we do for dropout).
3. There are quite a few `np` calls remaining in the modelling file, for example: https://github.com/huggingface/transformers/blob/843a6c05d99ad5da629d07576aa0eb83e7b17b07/src/transformers/models/big_bird/modeling_flax_big_bird.py#L579 and https://github.com/huggingface/transformers/blob/843a6c05d99ad5da629d07576aa0eb83e7b17b07/src/transformers/models/big_bird/modeling_flax_big_bird.py#L979-L980 Would you mind taking a look at these calls and seeing whether they can be changed to `jnp` calls so that we can correctly JIT the model?

2 is a bit fiddly, so here's a checklist you can follow to try and get it working this way:
* Create an RNG called `indices` in the attention function and assign it to the variable `indices_rng`. This is very similar to what we do for `dropout_rng` (more details in the comments below)
* Use the dict `rngs` to hold all your PRNG keys (`dropout` and `indices`) and use this to pass the keys to the Flax model (code suggestion bellow)
* If you do this correctly, you should be able to remove all the arguments `indices_prng_key` throughout the modelling file. The only place it should remain is **within** `FlaxBigBirdSelfAttention`

Bearnardd commented:
Hi @sanchit-gandhi! Thank you very much for this detailed review. It is really helpful since this is my first time working with JAX :). I will apply the changes during the weekend. Have a great day!

sanchit-gandhi commented:
Awesome, very glad to hear that the pointers were helpful 🤗 feel free to post here if you have any questions - it's a bit of a fiddly fix and I'm more than happy to help if you get stuck on anything!

There's actually a similar rng trick that we use in Flax BEIT:
https://github.com/huggingface/transformers/blob/b210c83a78022226ce48402cd67d8c8da7afbd8d/src/transformers/models/beit/modeling_flax_beit.py#L161

You can follow through the logic we employ with `"droppath"` and `droppath_rng` to see a working example of what we want to do here!

Bearnardd commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -978,24 +987,24 @@ def _bigbird_block_rand_mask(
                 if start > last:
                     start = last
-                    rand_attn[i - 1, :] = np.random.permutation(middle_seq[:start])[:r]
+                    rand_attn[i - 1, :] = jax.random.permutation(middle_seq[:start])[:r]
```
Not at all :)

Bearnardd commented:
Hi @sanchit-gandhi! Sorry for the late response but lately I was in the process of changing workplaces as well as on vacation so I have not checked github for a while :). I have implemented your comments but I have two follow up questions:

1) Should I remove all `numpy` calls in the modeling file even the ones like `np.zeros` or `np.arange` or only the ones related to the randomness?

2) I have some problems with  `indices_prng_key` for the scenario when `FlaxBigBirdBlockSparseAttention` is used but `deterministic=True` for which `indices_prng_key=None`. Since even though deterministic is set to False the random jax functions are still being called and in this case the provided `rng_key=None` which results in the error.

sanchit-gandhi commented:
Hey @Bearnardd! Awesome to see that you've picked-up this PR again!

1. Yes please! If you could replace all NumPy calls with their JAX equivalents that would be grand! This will keep all tensors on the accelerator device (GPU/TPU) rather than pulling them back to the host
2. In this case, could we add `if/else` logic that returns the correct attention mask in the deterministic case? E.g.
```python
if self.deterministic:
    # do the deterministic inference attention with no randomness
else:
    # do the stochastic training attention with jnp randomness
```
A similar logic is used in the Flax dropout module: https://flax.readthedocs.io/en/latest/_modules/flax/linen/stochastic.html#Dropout

Bearnardd commented:
Hi @sanchit-gandhi! I have replaced all NumPy calls but frankly I am not sure if I understand the second part correctly. Could you explain what do you mean by `deterministic inference attention` and where that `if/else` logic should be places?

sanchit-gandhi commented:
Hey @Bearnardd! Very cool! Do you mind pushing your changes to origin so that I can see the code? This might make it easier to help-out with the deterministic issue!

Essentially, we only want to do the random operations when we're training, not at inference time. During inference, we want everything to be deterministic. This is like dropout - we only do this during training and not inference, when we want to disable dropout and have all the nodes be active.

We can check if the model is deterministic through the attribute `self.determisitic` (like `self.training` in PyTorch). What we need to do is add some logic so that the random calls are only made if `self.deterministic=False` (training): we know we're in training mode and we want all of the randomness, so we activate all the random calls. 
_Else_ we're deterministic, then we don't want to do any of the randomness, e.g. skip all of it.

Bearnardd commented:
Hi @sanchit-gandhi! Sure I will push the changes around Friday since I am currently at a business trip and I do not have my personal laptop :/

Bearnardd committed:
modify indices_prng_key usage, add deterministic behaviour

Bearnardd commented:
Hi @sanchit-gandhi! I have pushed the changes.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -240,7 +238,6 @@ def __call__(self, input_ids, token_type_ids, position_ids, attention_mask, dete
         return hidden_states
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertSelfAttention with Bert->BigBird
```
This class is the same as FlaxBertSelfAttention with Bert->BigBird: don't think we should remove the `# Copied from...` statement

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -461,6 +458,10 @@ def __call__(
 
+        indices_prng_key = None
+        if not deterministic:
+            indices_prng_key = self.make_rng("indices")
```
Nice!

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -956,6 +971,7 @@ def _bigbird_block_rand_mask(
             from_block_size: int. size of block in from sequence.
             to_block_size: int. size of block in to sequence.
             num_rand_blocks: int. Number of random chunks per row.
+            indices_prng_key: jax.random.PRNGKey. PRNG key that is used to perform random jax operations
```
(nit!)
```suggestion
            indices_prng_key: jax.random.PRNGKey. PRNG key that is used to perform random jax operations.
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -966,9 +982,12 @@ def _bigbird_block_rand_mask(
             raise ValueError("Error the number of blocks needs to be same!")
+        rand_attn = jnp.zeros((from_seq_length // from_block_size - 2, num_rand_blocks), dtype=jnp.int32)
+        # deterministic nor randomness
+        if indices_prng_key is None:
```
I guess this `if` statement could also be:
```python
if deterministic:
```
Since `indices_prng_key` is None when we're deterministic. Happy with either! Think the `if deterministic` might be easier to understand logic wise?

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -978,25 +997,31 @@ def _bigbird_block_rand_mask(
             end = i
             if i == 1:
-                rand_attn[i - 1, :] = np.random.permutation(middle_seq[2:last])[:r]
+                rand_attn = rand_attn.at[i - 1].set(jax.random.permutation(indices_prng_key, middle_seq[2:last])[:r])
```
Nice! We could split this over two lines for readability:
```suggestion
                seq_values = jax.random.permutation(indices_prng_key, middle_seq[2:last])[:r]
                rand_attn = rand_attn.at[i - 1].set(seq_values)
```

Bearnardd replied:
Yeah, we definitely should :)

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -978,25 +997,31 @@ def _bigbird_block_rand_mask(
-                    )[:r]
+                    rand_attn = rand_attn.at[i - 1].set(
+                        jax.random.permutation(
+                            indices_prng_key, jnp.concatenate((middle_seq[:start], middle_seq[end + 1 : last]))
```
Especially when we have the concat we should definitely split over two lines for readability ✅

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1026,6 +1052,7 @@ def _bigbird_block_rand_mask_with_head(
             num_heads: int. total number of heads.
             plan_from_length: list. plan from length where num_random_blocks are choosen from.
             plan_num_rand_blocks: list. number of rand blocks within the plan.
+            indices_prng_key: jax.random.PRNGKey. PRNG key that is used to perform random jax operations
```
Same again here just for consistency
```suggestion
            indices_prng_key: jax.random.PRNGKey. PRNG key that is used to perform random jax operations.
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1048,15 +1075,23 @@ def _bigbird_block_rand_mask_with_head(
         # Random Attention adjacency list
         rand_attn = [
-            np.zeros((num_blocks, np.sum(plan_num_rand_blocks[: max_plan_idx + 1])), dtype=np.int32)
+            jnp.zeros((num_blocks, sum(plan_num_rand_blocks[: max_plan_idx + 1])), dtype=jnp.int32)
```
Would use a `jnp.sum` on jax arrays as it should be faster!
```suggestion
            jnp.zeros((num_blocks, jnp.sum(plan_num_rand_blocks[: max_plan_idx + 1])), dtype=jnp.int32)
```

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1048,15 +1075,23 @@ def _bigbird_block_rand_mask_with_head(
         ]
 
+        # deterministic
+        if indices_prng_key is None:
```
Why do we do this now if running in deterministic mode? (just for my understanding!)

Bearnardd replied:
We want to return zeros but to match the expected output shape of this method we have to match the slicing applied to the `rand_attn` tensor and the end of this method for non deterministic case.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1067,11 +1102,11 @@ def _bigbird_block_rand_mask_with_head(
                 if plan_num_rand_blocks[plan_idx] > 0:
-                    rnd_r_cnt = int(np.sum(plan_num_rand_blocks[:plan_idx]))
-                    curr_r_cnt = int(np.sum(plan_num_rand_blocks[: plan_idx + 1]))
+                    rnd_r_cnt = int(sum(plan_num_rand_blocks[:plan_idx]))
```
Think jnp.sum's here would be better too

Bearnardd replied:
@sanchit-gandhi  I have used that in the first place, but while using it for example in line `1082` during `rand_attn` initialization, it requires casting `plan_num_rand_blocks` to `jnp.array` which results in the error that I wasn't able to fix:

```
jax._src.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=1/0)>
           While tracing the function init_weights at /home/bearnard/Dev/transformers/src/transformers/models/big_bird/modeling_flax_big_bird.py:1640 for eval_shape, this value became a tracer due to JAX operations on these lines:

            operation a:i32[2] = convert_element_type[new_dtype=int32 weak_type=False] b
               from line /home/bearnard/Dev/transformers/src/transformers/models/big_bird/modeling_flax_big_bird.py:1083 (<listcomp>)
           See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
```

sanchit-gandhi replied:
That's weird! Went through quite a bit of the JAX code to see what they use, seems like `sum()` and `jnp.sum()` are both fine, so if `sum()` works lets stick with that.

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1179,14 +1224,14 @@ def _get_single_block_row_attention(
             illegal_blocks.append(1)
 
-        selected_random_blokcs = []
+        selected_random_blocks = []
```
Good catch!

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1080,6 +1115,10 @@ def _bigbird_block_rand_mask_with_head(
                                 global_block_right=global_block_right,
+                                indices_prng_key=indices_prng_key,
+                            )
+                            rand_attn[h] = (
```
Nice! Handled these very cleanly

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1324,7 +1369,6 @@ def setup(self):
         if self.config.add_cross_attention:
             self.crossattention = FlaxBigBirdAttention(self.config, causal=False, dtype=self.dtype)
 
-    # Copied from transformers.models.bert.modeling_flax_bert.FlaxBertLayer.__call__ with Bert->BigBird
```
We should make sure these all go back in for the final PR

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1606,8 +1647,8 @@ def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: Froz
 
-        params_rng, dropout_rng = jax.random.split(rng)
-        rngs = {"params": params_rng, "dropout": dropout_rng}
+        params_rng, dropout_rng, indices_rng = jax.random.split(rng, num=3)
```
Nice!

sanchit-gandhi reviewed (approved):
Very nice work @Bearnardd! The PR is looking really great 👌 Reading through the code, the JAX functionality all looks good to me. Think it's mainly some style formatting before we can get this merged! I can't see what tests are passing/failing at the moment because there's a `pdb` debugging statement in the modelling code that is holding-up the CI:
https://github.com/huggingface/transformers/blob/fa55da25a904c121868928e16968731ccd10f647/src/transformers/models/big_bird/modeling_flax_big_bird.py#L1850

Are the tests passing locally for you otherwise?

Bearnardd committed:
update style

Bearnardd committed:
remove unused import

Bearnardd committed:
remove copy statement since classes are not identical

Bearnardd committed:
Merge branch 'main' of https://github.com/huggingface/transformers into fix_bigbird_random_attention

Bearnardd committed:
remove numpy import

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1386,7 +1432,6 @@ def setup(self):
                 for i in range(self.config.num_hidden_layers)
             ]
 
-    # Copied from transformers.models.bert.modeling_flax_bert.FlaxBertLayerCollection.__call__ with Bert->BigBird
```
We can keep this in

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1451,7 +1496,6 @@ def __call__(
         )
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertEncoder with Bert->BigBird
```
Same here :)

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1507,11 +1551,10 @@ def __call__(self, hidden_states):
         return self.LayerNorm(hidden_states)
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertLMPredictionHead with Bert->BigBird
```
Also here! (just the type hint that changes)

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1637,7 +1685,6 @@ def init_weights(self, rng: jax.random.PRNGKey, input_shape: Tuple, params: Froz
         else:
             return random_params
 
-    # Copied from transformers.models.bart.modeling_flax_bart.FlaxBartDecoderPreTrainedModel.init_cache
```
I think this is the same no?

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1834,7 +1889,6 @@ class FlaxBigBirdModel(FlaxBigBirdPreTrainedModel):
 append_call_sample_docstring(FlaxBigBirdModel, _CHECKPOINT_FOR_DOC, FlaxBaseModelOutputWithPooling, _CONFIG_FOR_DOC)
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertForPreTrainingModule with Bert->BigBird
```
And here too

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1936,7 +1990,6 @@ class FlaxBigBirdForPreTraining(FlaxBigBirdPreTrainedModel):
 )
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertForMaskedLMModule with Bert->BigBird
```
Also here :)

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -2099,7 +2152,6 @@ class FlaxBigBirdForSequenceClassification(FlaxBigBirdPreTrainedModel):
 )
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertForMultipleChoiceModule with Bert->BigBird
```
This one as well

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -2198,7 +2250,6 @@ def __init__(
 )
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertForTokenClassificationModule with Bert->BigBird
```
Here too!

sanchit-gandhi reviewed (approved):
Awesome work @Bearnardd! Just a few `# Copied from...` statements that we can put back in, but the heavy JAX work looks good to me! I've labelled the ones you can put in for convinience below.

Bearnardd committed:
Merge branch 'main' of https://github.com/huggingface/transformers into fix_bigbird_random_attention

Bearnardd committed:
revert removing copied from statements

Bearnardd commented:
Hi @sanchit-gandhi all copied from statements are back :)

Bearnardd committed:
make style from copied

Bearnardd committed:
remove copied from statement

sanchit-gandhi commented on src/transformers/models/big_bird/modeling_flax_big_bird.py:
```diff
@@ -1553,7 +1553,6 @@ def __call__(self, hidden_states):
         return self.LayerNorm(hidden_states)
 
 
-# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertLMPredictionHead with Bert->BigBird
```
Hey @Bearnardd! Great, thanks! We can do a sneaky trick here where we do:
```
# Copied from transformers.models.bert.modeling_flax_bert.FlaxBertLMPredictionHead with Bert->BigBird, np->jnp
```
to copy the class but change the type annotation.

Otherwise this PR is ready for a final review! Super work on getting it finished - it was quite a tricky PR handled very well with some new JAX PRNG logic!

Bearnardd replied:
Wow! I had no idea this is possible. Nice  one :)

sanchit-gandhi replied:
Ah yes good spot! Hmm that's interesting, sounds like a bug in Flax BERT! Feel free to open a PR to fix it, otherwise I can do it after we merge this PR :)

Bearnardd committed:
update copied from statement to include only np.ndarry

amyeroberts reviewed (approved):
Very nice PR - thanks for all the work fixing it! 

Only small comments. Main one is about the status of the equivalence tests - if it's not possible to have them passing now, could you add the `unittest.skip` decorators and open an issue so this doesn't get lost?

Bearnardd committed:
Merge branch 'main' of https://github.com/huggingface/transformers into fix_bigbird_random_attention

Bearnardd committed:
add deterministic args, unittestskip equivalence tests

Bearnardd commented:
Hi @amyeroberts  @sanchit-gandhi! I changed if checking to `deterministic` and added `unittestskip` for equivalence tests. Probably around weekend I will create a issue regarding bug in Pytorch's implementation as well as PR fix. Nevertheless I guess this PR is ready to be merged.

sanchit-gandhi reviewed (approved):
Thanks for making the final changes @Bearnardd! Good for me - will let @amyeroberts confirm she's happy before merging!

amyeroberts commented:
Yep - it all looks good to me. Thanks again for this contribution, @Bearnardd!
